staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -667,7 +670,16 @@ func (c *Cacher) dispatchEvent(event *watchCacheEvent) {
 	c.startDispatching(event)
 
 	// Since add() can block, we explicitly add when cacher is unlocked.
+
+	// Dispatching event in nonblocking way firstly, which make faster watchers
+	// not be blocked by slower ones.
+	c.watchersSlower = c.watchersSlower[:0]
 	for _, watcher := range c.watchersBuffer {
+		if !watcher.nonblockingAdd(event) {
+			c.watchersSlower = append(c.watchersSlower, watcher)
+		}
+	}
+	for _, watcher := range c.watchersSlower {
```

Comment:
I think we can safe more if we slightly change it. What i suggest is the following:
```
if len(watchersSlower) > 0 {
  // ... move the comment from below here
  startTime = time.Now()
  timeout := budget.takeAvailable()
  timer.Reset(timeout)
  for _, watcher := range c.watchersSlower {
    watcher.add(event, timer)
  }
  if !c.timer.Stop() {
    <-c.timer.C
  }
  budget.returnUnused(timeout - time.Since(startTime)
}
```

[with all the comments moved from add]

If we do that, we basically reduce number of operations on budget, operations on timers etc.